File: apps/server/src/ports/driven/services/shopify-client.ts
// src/ports/driven/services/shopify-client.ts
//
// Port for Shopify integration.
// Speaks in domain terms (PilotProduct, ShopifyProductId).
// The adapter handles the translation to Shopify's API format.

import { Context, Data } from 'effect'
import type { Effect } from 'effect/Effect'

import type { PilotProduct, ShopifyProductId } from '../../../domain/pilot'

// ============================================
// SHOPIFY CLIENT ERROR
// ============================================

export class ShopifyClientError extends Data.TaggedError('ShopifyClientError')<{
  readonly operation: string
  readonly cause: unknown
}> {}

// ============================================
// SHOPIFY CLIENT
// ============================================

/**
 * Port for Shopify product synchronization.
 *
 * This interface speaks in domain terms:
 * - Input: PilotProduct (domain aggregate)
 * - Output: ShopifyProductId (domain value object)
 *
 * The adapter implementation handles:
 * - Mapping domain → Shopify API format
 * - API calls
 * - Error handling
 */
export interface ShopifyClientService {
  readonly syncProduct: (product: PilotProduct) => Effect<ShopifyProductId, ShopifyClientError>

  readonly archiveProduct: (shopifyProductId: ShopifyProductId) => Effect<void, ShopifyClientError>
}

export class ShopifyClient extends Context.Tag('ShopifyClient')<
  ShopifyClient,
  ShopifyClientService
>() {}

File: apps/server-java/src/main/java/com/maisonamane/port/driven/service/ShopifyClient.java
package com.maisonamane.port.driven.service;

import com.maisonamane.domain.pilot.aggregate.PilotProduct;
import com.maisonamane.domain.pilot.valueobject.ShopifyProductId;
import io.vavr.control.Either;

/**
 * Port for Shopify API client. Maps Effect-TS Context.Tag<ShopifyClient> to Java interface.
 */
public interface ShopifyClient {

    sealed interface ShopifyError {
        String message();

        record ApiError(String message, int statusCode) implements ShopifyError {}

        record ValidationError(String message, java.util.List<String> fields) implements ShopifyError {}

        record NetworkError(String message, Throwable cause) implements ShopifyError {}
    }

    Either<ShopifyError, ShopifyProductId> syncProduct(PilotProduct product);
}
